proposals/0000-slurp.rst

```diff
@@ -0,0 +1,454 @@
+uniformly. But how do we build them? This supposes a common framework
+for specifying how to build a package. We already have such a framework:
+Cabal-the-spec. We furthermore already have an implementation of this
+specification: Cabal-the-library and the various ``Setup.hs`` build
+files that authors have been shipping with their packages.
+
+So to answer our original question, the minimum we need to *share* is:
+
+-  A centralized understanding of who "owns" what package name and where
+   to go to find all package versions the maintainer has published under
+   this name.
+-  A central registry that implements this understanding: this is SLURP.
+-  A common format for package metadata (i.e. Cabal) and a common way to
+   build packages.
+-  A standard API for performing (1), (2) and (3) above.
+
+Given a central registry to locate packages (SLURP), any build tool or
+package infrastructure can query which server a package is hosted on.
+Then, provided that server implements a standard API, the build tool can
+query available versions and download source distributions.
+
+4 The SLURP API
+---------------
+
+The central registry, SLURP, is a new service. Its purpose is to allow
+clients claim a package name, to list all available packages, and to
+find where a package is hosted. It implements the following API:
+
+-  ``GET /packages`` - returns a list of all open source packages in the
+   Haskell universe.
+-  ``PUT /package/:pkgname`` - register ownership of a new package name.
+   Map that name to some URL that you own. Example:
+   ``{"name": "mypackage", "location": "https://myserver.com/package/mypackage" }``
+-  ``GET /package/:pkgname`` - redirects to URL provided by ``PUT``.
+
+No authentication or user accounts required to make requests to either
+of those endpoints.
+
+Example package list returned as the response of a ``GET /packages``
+request:
+::
+
+    { "packages":
+      [ {"name": "bytestring", "location": "https://hackage.haskell.org/package/bytestring"}
+      , {"name": "text", "location": "https://hackage.haskell.org/package/text"}
+      , {"name": "Cabal", "location": "https://hackage.haskell.org/package/Cabal"}
+      , {"name": "conduit", "location": "https://stackage.org/package/conduit" }
+      , {"name": "yesod", "location": "https://stackage.org/package/yesod" }
+      , {"name": "mypackage", "location": "https://myserver.com/package/mypackage" }
+      ...
+      ]
+    }
+
+You can think of SLURP as an *authoritative* URL-shortening service for
+packages. Authoritative means that which package is generally accepted
+to have which name depends on the content of the SLURP registry and only
+that.
+
+5 The package API
+-----------------
+
+The target location for each registered package must support at least
+the following subset of the Hackage API:
+
+-  ``GET /package/:pkgname/preferred`` returns a JSON structure listing
+   all versions.
+-  ``GET /package/:pkgname/:tarball.tar.gz`` where tarball is a
```

Comment:
While I disagree with the proposal as a whole, I don't think that tying ourselves to tgz is a good idea.